sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -373,13 +397,70 @@ public void validate(PipelineOptions options) {
 
     boolean fixedSharding = getComputeNumShards() != null || getNumShardsProvider() != null;
     PCollection<List<FileResult<DestinationT>>> tempFileResults;
-    if (fixedSharding) {
+    if (getStageFilesDirectly()) {
+      checkArgument(
+          input.getWindowingStrategy().getWindowFn() instanceof FixedWindows,
+          "Must be a fixed window to promote window to key.");
```

Comment:
Can this check be relaxed/generalized?

Reply:
I definitely can be relaxed to accept non-merging windows.  We could accept merging windows as well but the behavior might less consistent because we won't run window merging before the groupbykey and thus only exact window matches would possibly stage directly. It might be better to just disallow for that reason.

If we move this behavior to  the non-fixed sharding path using GroupIntoBatches, that also doesn't currently support windowing/triggering either:
https://issues.apache.org/jira/browse/BEAM-12040

I kind of agree with https://issues.apache.org/jira/browse/BEAM-4604 that windowing and triggering is a bit confusing when it is applied to composite transforms such as FileIO.  Unless you know the implementation does or does not use GroupByKey you don't know if the windowing/triggering will be have any effect.